### PR TITLE
Feature/darkmode

### PR DESCRIPTION
**Pull request by martindevtum**
### Checklist
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes
- [x] I translated all the newly inserted strings into German and English

### Motivation and Context
Darkmodes and themings and general can provide a way to make working on Apollon more enjoyable. A light on dark theme for example can provide a way to reduce strain on the eye when working for. long periods of times. A Darkmode is also a highly demanded feature (https://github.com/ls1intum/Artemis/issues/4404). 
Hence we want to create a Darkmode to use in Apollon and Artemis.
Thus we here define a solution for a Darkmode in Apollon with an underlying theming system relying on css root variables for colors.
Those root variables can be injected by consuming Applications, making Apollons theming highly flexible.

#### Reducing Theming
Firstly in order to make theming easily adjustable i reduced our prior existing theming solution as much as possible. 
I here reduced  the prior theming from containing the following 14 colors that were not all in use:
`[primary, secondary, warningYellow, white, gray100, gray200, gray300, gray400, gray500, gray600, gray700, gray800, gray900, black]`

to a new theming system consisting of only 8 colors with more generic names:
`
  --apollon-primary
  --apollon-primary-contrast
  --apollon-secondary
  --apollon-warning-yellow
  --apollon-background
  --apollon-background-variant
  --apollon-gray
  --apollon-gray-variant
`

These colors are expected to be present as :root {} color variables. If none exist, we fallback to the light-themed color sheme.

#### Bringing Theming into the components
In order to theme our components, we already used styled-components.
I created styled-components in `themedComponents.ts` for our svgs, these components make sure to provide the current theming. Instead of using `<rect \>` to create an svg rectangle, one can now use <ThemedRect \> and the rect would now be automatically themed.

#### Create a theming
To test this, i created a dark theme for the apollon example app.
The theming is defined in `public/themings.json` and get applied by the added Theming buttons in the sidebar.
The exact definition of the theming can well be up to debate, but with this infrastructure it can also be easily adjusted.

### Steps for Testing
start the application locally and test it for colors.

1.  yarn install && yarn start 
2. switch the theming by using the buttons in the sidebar on the right

### Screenshots

lightmode-example
darkmode-example

### CODE REVIEW

**Comment by Hialus** [comment]
I am curious how you came up with the theme for the dark color palette.
Did you use an existing color palette, or did you create your own using some guide?

Since this is a very subjective change that will also affect Artemis, we should consider creating a poll with different dark color palettes and take the most popular one. (This could either be a dev only poll or a public poll like it was done with the new Artemis logo)
For example, I personally feel like the background color looks wrong. For me, it is a bit too heavy in the green.

**Comment by martindevtum** [comment]
> I am curious how you came up with the theme for the dark color palette. Did you use an existing color palette, or did you create your own using some guide?
> 
> Since this is a very subjective change that will also affect Artemis, we should consider creating a poll with different dark color palettes and take the most popular one. (This could either be a dev only poll or a public poll like it was done with the new Artemis logo) For example, I personally feel like the background color looks wrong. For me, it is a bit too heavy in the green.

As I have already mentioned in the description, the exact color palette can still be subject of change, and I chose those colors to exemplify how my changes work.
The thing is, that currently, nothing would change for Artemis (Of course we would need to test that. in an Artemis PR), as the colors if none are defined would fallback to the already existing light-mode. 

I love the idea of creating a Poll. 
The question however is wether we should do that for this PR or wait until more progress is done in the Artemis Darkmode.
I think the different color palettes should be defined not only for Apollon, but for Artemis and Apollon together, to make sure they harmonize.

**Comment by Hialus** [comment]
> The question however is whether we should do that for this PR or wait until more progress is done in the Artemis Darkmode.
I think the different color palettes should be defined not only for Apollon, but for Artemis and Apollon together, to make sure they harmonize.

I guess first implementing the functionality and taking care of the proper theme later could also work. I just personally would do the poll first, as this reduced the workload for you. Cause even a simple theming update in Apollon requires a follow-up PR in Artemis and Apollon-Standalone

But doing the functionality first and taking care of the proper Apollon theme later would have an advantage: We could implement a bunch of different dark themes in Artemis and then use that for the poll. It would make a lot of sense to do it this way.
My only question now would be this: Would Artemis be able to override the Apollon theme or does it merely set it? If its the latter we would also need to add the different dark themes in Apollon (Obviously this could be done in a separate PR)

**Comment by martindevtum** [comment]
> > The question however is whether we should do that for this PR or wait until more progress is done in the Artemis Darkmode.
> > I think the different color palettes should be defined not only for Apollon, but for Artemis and Apollon together, to make sure they harmonize.
> 
> I guess first implementing the functionality and taking care of the proper theme later could also work. I just personally would do the poll first, as this reduced the workload for you. Cause even a simple theming update in Apollon requires a follow-up PR in Artemis and Apollon-Standalone
> 
> But doing the functionality first and taking care of the proper Apollon theme later would have an advantage: We could implement a bunch of different dark themes in Artemis and then use that for the poll. It would make a lot of sense to do it this way. My only question now would be this: Would Artemis be able to override the Apollon theme or does it merely set it? If its the latter we would also need to add the different dark themes in Apollon (Obviously this could be done in a separate PR)


So far the plan is, that consuming projects such as Artemis or Apollon_standalone would be setting the themes via root css variables. So they can inject a theming into apollon. Providing pre-defined themings for Apollon from within Apollon is something I haven't put much thought in to be honest, as it is a library that can and should be adjusted by consuming applications to match their themings. In reality it wouldn't be hard to achieve, but I would question the necessity for it. 
This PR just builds the infrastructure required to make Apollon adjustable.  
However I would also wait with merging this PR until more progress is done in the Artemis theming, as this could introduce new unforseen requirements and problems. I will create a PR in Apollon_standalone introducing a darkmode, soon. However this is still blocked by ( https://github.com/ls1intum/Artemis/pull/4827 )

**Comment by Hialus** [comment]
Great. Thanks for clarifying!
So the theme in Apollon itself would be more of an example.
> However I would also wait with merging this PR until more progress is done in the Artemis theming, as this could introduce new unforseen requirements and problems.

Makes sense 👍🏻

**Comment by bassner** [comment]
@Hialus @martindevtum As long as Apollon defaults to the current theme and allows overrides using CSS variables, everything is fine. Adding a dark mode / theming to Artemis will be part of my thesis (it's more complex than one might think on first sight), and I can see that I fix any upcoming issues throughout my time working on it. It's awesome that this PR provide some groundwork to properly include Apollon into a themed UI! Thank you :)

**Comment by minrows** [comment]
Hi @martindevtum , If you're still working on this PR, could you please convert it to draft?

**Comment by minrows** [comment]
I did test in my local machine, the color switching is working as expected. 
But there are a few unexpected issues. One of them being the diagram of package is not rendered properly. Could you make sure its not happening in yours?
image

**Comment by martindevtum** [comment]
> I did test in my local machine, the color switching is working as expected. But there are a few unexpected issues. One of them being the diagram of package is not rendered properly. Could you make sure its not happening in yours? image

Hey Swornim, thanks, great catch!
A path was accidentally converted into a rect. I checked al the components once again and couldn't find any more of those issues.

**Comment by martindevtum** [comment]
> Hi @martindevtum , If you're still working on this PR, could you please convert it to draft?

This PR is ready for review.

**Comment by minrows** [comment]
> > Hi @martindevtum , If you're still working on this PR, could you please convert it to draft?
> 
> This PR is ready for review.

Could you please make sure that all the build checks are passing?

**Comment by martindevtum** [comment]
> > > Hi @martindevtum , If you're still working on this PR, could you please convert it to draft?
> > 
> > 
> > This PR is ready for review.
> 
> Could you please make sure that all the build checks are passing?

@minrows all passing.

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@ls1intum/apollon",
-  "version": "2.9.1",
+  "version": "2.9.1-beta.0",
```
Comment (minrows) [inline package.json:3]: Any specific reason for the version change?

**Review by minrows — COMMENTED** [review]
Functionality wise, its working great, tested locally. Few observations though:

- Actor in dark mode could be made more visually appealing by making its body white too.
- Version is set to beta, not quite sure about the purpose of it.
- Canvas now has a margin on top and left sides. Was that intentional? I personally like it this way (with margins)
- And as mentioned in the comments before, we can always work on the color selection. 

PS: Great work @martindevtum . 👍

File: package.json (hunk 1 above)
Comment (martindevtum) [reply package.json:3]: I created a beta release from this branch.  Will be adjusted

**Review by minrows — APPROVED** [review]

**Pull request merged**